src/redisgears_python.c

```diff
@@ -2093,6 +2094,309 @@ static PyObject* createTensorFromValues(PyObject *cls, PyObject *args){
     return NULL;
 }
 
+static PyObject* setTensorInKey(PyObject *cls, PyObject *args) {
+    verifyRedisAILoaded();
+    // The function returns Py_None in any case, prepare it.
+    Py_INCREF(Py_None);
+    // Input validation: 2 arguments. args[0]: string. args[1]: tensor
+    if(PyTuple_Size(args) != 2){
+        PyErr_SetString(GearsError, "Wrong number of arguments given to setTensorInKey");
+        return Py_None;
+    }
+    PyObject* keyName = PyTuple_GetItem(args, 0);
+    if(!PyUnicode_Check(keyName)){
+        PyErr_SetString(GearsError, "key name argument must be a string");
+        return Py_None;
+    }
+
+    
+    PyTensor* pyt = (PyTensor*)PyTuple_GetItem(args, 1);
+    if(!PyObject_IsInstance((PyObject*)pyt, (PyObject*)&PyTensorType)){
+        PyErr_SetString(GearsError, "Given argument is not of type PyTensor");
+        return Py_None;
+    }
+
+    RedisModuleType *ai_tensorType = RedisAI_TensorRedisType();
+    size_t len;
+    const char* keyNameCStr = PyUnicode_AsUTF8AndSize(keyName, &len);
+
+    if(Cluster_IsClusterMode()){
+        const char* keyNodeId = Cluster_GetNodeIdByKey(keyNameCStr);
+        if(!Cluster_IsMyId(keyNodeId)){
+            PyErr_SetString(GearsError, "Given key is not in the current shard");
+            Py_INCREF(Py_None);
+            return Py_None;
+        }
+    }
+
+    RedisModuleCtx* rctx = RedisModule_GetThreadSafeContext(NULL);
+    RedisModuleString *keyNameStr = RedisModule_CreateString(rctx, keyNameCStr, len);
+    LockHandler_Acquire(rctx);
+    RedisModuleKey *key = RedisModule_OpenKey(rctx,keyNameStr, REDISMODULE_WRITE);
+    // Check if the key is empty or 
+    bool empty = RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_EMPTY;
+    if(!empty && (RedisModule_ModuleTypeGetType(key) != ai_tensorType)) {
+        RedisModule_CloseKey(key);
+        LockHandler_Release(rctx);
+        RedisModule_FreeString(rctx, keyNameStr);
+        RedisModule_FreeThreadSafeContext(rctx);
+        PyErr_SetString(GearsError, "ERR Key is already set with non tensor values.");
+        return Py_None;
+    }
+    // Set a shllow copy of the tensor.
+    RedisModule_ModuleTypeSetValue(key, ai_tensorType, RedisAI_TensorGetShallowCopy(pyt->t));
+    RedisModule_CloseKey(key);
+    LockHandler_Release(rctx);
+    RedisModule_FreeString(rctx, keyNameStr);
+    RedisModule_FreeThreadSafeContext(rctx);
+    return Py_None;
+}
+
+static PyObject* msetTensorsInKeyspace(PyObject *cls, PyObject *args) {
```

Comment:
Why you need also mset and setTensorInKey. Isn't mset cover set?

Reply:
I would think about this all function design again. Lots of things can be allocated on the stack instead of array_new, I suggest to have a goto error and clean the memory there...

Reply:
I don't mind removing the `set`, I thought about `mset` after I had `set` implemented, so I didn't want to delete

also, all lists are now stack allocated arrays